skimage/_shared/utils.py

```diff
@@ -243,7 +243,8 @@ def convert_to_float(image, preserve_range):
         Transformed version of the input.
     """
     if preserve_range:
-        image = image.astype(np.double)
+        if image.dtype.char not in 'df':
```

Comment:
```suggestion
        if not np.issubdtype(image.dtype, np.floating):
```

```python
>>> import numpy as np
>>> np.issubdtype(np.float32, np.floating)
True
>>> np.issubdtype(np.float64, np.floating)
True
>>> np.issubdtype(np.float16, np.floating)
True
>>> np.issubdtype(np.uint8, np.floating)
False
```

Reply:
That's definitely more readable ;-). I excluded `float16` and `float128` because I though that these types are not supported by `skimage`.

Reply:
This will cause a bug since `float16` is not supported by `cython.floating` (only `float` and `double`) :-(

```python
>>> a = np.ones((5, 5), dtype=np.float16)
>>> resize(a, (10, 10))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "skimage/transform/_warps_cy.pyx", line 70, in skimage.transform._warps_cy._warp_fast
    def _warp_fast(floating[:, :] image, floating[:, :] H, output_shape=None,
ValueError: Does not understand character buffer dtype format string ('e')
```
* We can use
  ```python
  if image.dtype not in [float, np.float, np.double, np.float32, np.float64]:
  ```
  but I am not sure that it is more readable...
* The other option is to keep the original test (`if image.dtype.char not in 'df'`) and to add a comment in the code.

What do you think?

Reply:
Just ignore float16, the whole library is broken with it.

Reply:
All right, I will have to ignore `np.float128` also:
```python
>>> a = ones((5, 5), dtype=np.float128)
>>> resize(a, (10, 10))
ValueError: Buffer dtype mismatch, expected 'double' but got 'long double'
```
This is because `cython.floating` only supports single and double precision floating points.
Do you prefer `if not (np.issubdtype(image.dtype, np.floating) or image.dtype in [np.float16, np.float128])` instead of `if image.dtype.char not in 'df'`?

Reply:
I once wrote a fused numerics file. You can add things there.

That said, neither of those two datatypes are efficient in hardware, nor are they well tested. 

If you want to, you can add to the fused numerics file maybe. Not exactly sure what will happen to the rest of the library.

Reply:
Honestly, you could just add a comment. That said, you could also add a Todo, and raise an issue that float16 and float128 have yet undefined behaviours in the img_as routines.

Reply:
I already tried some explorations with `skimage._shared.fused_numerics.np_floats`, but `numpy` doesn't declare `float16_t` and `float128_t` for `cimport` :-(

Reply:
yeah, float16 and float128 are pretty badly supported in general. Adding their definitions to the cimport could be added as an issue upstream.